src/internet/subsonic/subsonicdynamicplaylist.cpp

```diff
@@ -0,0 +1,259 @@
+  int task_id = service->app_->task_manager()->StartTask(tr("Fetching Playlist Items"));
+  QUrl url = service->BuildRequestUrl("GetAlbumList");
+  QNetworkAccessManager network;
+
+  switch (stat_) {
+    case QueryStat::QueryStat_Newest:
+      url.addQueryItem("type","newest");
+      break;
+    case QueryStat::QueryStat_Highest:
+      url.addQueryItem("type","highest");
+      break;
+    case QueryStat::QueryStat_Frequent:
+      url.addQueryItem("type","frequent");
+      break;
+    case QueryStat::QueryStat_Recent:
+      url.addQueryItem("type","recent");
+      break;
+  case QueryStat::QueryStat_Starred:
+      url.addQueryItem("type","starred");
+      break;
+    case QueryStat::QueryStat_Random:
+      url.addQueryItem("type","random");
+      break;
+  }
+
+  if (count > 500) count = 500;  // 500 limit per subsonic api
+  if (count != 10) { // 10 is default
+    url.addQueryItem("size",QString::number(count));
+  }
+
+  if (offset_ != 0) { // 0 is default
+    url.addQueryItem("offset",QString::number(offset_));
+  }
+
+  PlaylistItemList items;
+
+  QNetworkReply* reply = Send(network,url,service->usesslv3_);
+
+  // wait for reply
+  {
+    QEventLoop loop;
+    connect(reply, SIGNAL(finished()), &loop, SLOT(quit()));
+    loop.exec();
+  }
+
+  if (reply->error() != QNetworkReply::NoError) {
+    qLog(Warning) << "HTTP error returned from Subsonic:" << reply->errorString()
+                  << ", url:" << url.toString();
+    service->app_->task_manager()->SetTaskFinished(task_id);
+    return items; // empty
+  }
+
+  reply->deleteLater();
+  QXmlStreamReader reader(reply);
+  reader.readNextStartElement();
+  if (reader.name() != "subsonic-response") {
+    qLog(Warning) << "Not a subsonic-response, aboring playlist fetch";
+    service->app_->task_manager()->SetTaskFinished(task_id);
+    return items;
+  }
+
+  if (reader.attributes().value("status") != "ok") {
+    reader.readNextStartElement();
+    int error = reader.attributes().value("code").toString().toInt();
+    qLog(Warning) << "An error occured fetching data.  Code: "<<error<<" Message: "<<reader.attributes().value("message").toString();
+  }
+
+  reader.readNextStartElement();
+  if (reader.name() != "albumList") {
+    qLog(Warning) << "albumList tag expected.  Aboring playlist fetch";
+    service->app_->task_manager()->SetTaskFinished(task_id);
+    return items;
+  }
+
+  while (reader.readNextStartElement()) {
+    if (reader.name() != "album") {
+      qLog(Warning) << "album tag expected. Aboring playlist fetch";
+      service->app_->task_manager()->SetTaskFinished(task_id);
+      return items;
+    }
+
+    qLog(Debug) << "Getting album: "<<reader.attributes().value("album").toString();
+    GetAlbum(service,items,reader.attributes().value("id").toString(),network,service->usesslv3_);
+
+    reader.skipCurrentElement();
+  }
+  offset_+=count;
+  service->app_->task_manager()->SetTaskFinished(task_id);
```

Comment:
Might be an idea to write a helper class or use `BOOST_SCOPE_EXIT` to make sure you don't forget to call `SetTaskFinished()` before every return.